contracts/DivviRegistry.sol

```diff
@@ -159,6 +197,91 @@ contract DivviRegistry is
     emit RewardsAgreementRegistered(msg.sender, rewardsConsumer);
   }
 
+  /**
+   * @notice Register multiple referrals in a single transaction
+   * @dev Requires REFERRAL_REGISTRAR_ROLE
+   * @param referrals Array of referral data to register
+   */
+  function batchRegisterReferral(
+    ReferralData[] calldata referrals
+  ) external onlyRole(REFERRAL_REGISTRAR_ROLE) {
+    for (uint256 i = 0; i < referrals.length; i++) {
+      ReferralData calldata referral = referrals[i];
+
+      // Process the referral and get the status
+      ReferralStatus status = _registerReferral(
+        referral.user,
+        referral.rewardsProvider,
+        referral.rewardsConsumer,
+        referral.txHash,
+        referral.chainId
+      );
+
+      // Emit appropriate event based on status
+      if (status == ReferralStatus.SUCCESS) {
+        emit ReferralRegistered(
+          referral.user,
+          referral.rewardsProvider,
+          referral.rewardsConsumer,
+          referral.chainId,
+          referral.txHash
+        );
+      } else {
+        emit ReferralSkipped(
+          referral.user,
+          referral.rewardsProvider,
+          referral.rewardsConsumer,
+          referral.chainId,
+          referral.txHash,
+          status
+        );
+      }
+    }
+  }
+
+  /**
+   * @notice Register a user as being referred to a rewards agreement
+   * @dev Internal function that returns status instead of emitting events
+   * @param user The address of the user being referred
+   * @param rewardsProvider The address of the rewards provider entity
+   * @param rewardsConsumer The address of the rewards consumer entity
+   * @param txHash The hash of the transaction that initiated the referral
+   * @param chainId The ID of the blockchain where the referral transaction occurred
+   * @return status The status of the referral registration
+   */
+  function _registerReferral(
```

Comment:
Hmm, I notice that `txHash` and `chainId` are totally unused throughout the contract besides being emitted in events... is it necessary to provide them to the contract at all, since our registrar address is the only address permitted to submit referrals, and we'll be verifying transactions ourselves before submitting referrals to the contract?

Reply:
Yes I think this is good to have as an audit trace.
"ok the referral verifier is centralized for now, but anyone can externally see it correctly attributes referrals, by having a link to the original TX containing the referral data tag".